File: frontend/src/api/benefitsV2.ts
/**
 * API client for Benefits V2 endpoints
 * Provides typed access to the unified benefits catalog API
 */

import { api } from './client';
import type { CitizenProfile, Benefit, EligibilityRule, EstimatedValue } from '../engine/types';

// ========== API RESPONSE TYPES ==========

export interface BenefitSummaryAPI {
  id: string;
  name: string;
  shortDescription: string;
  scope: 'federal' | 'state' | 'municipal' | 'sectoral';
  state?: string;
  municipalityIbge?: string;
  estimatedValue?: EstimatedValue;
  status: string;
  icon?: string;
  category?: string;
}

export interface BenefitDetailAPI extends BenefitSummaryAPI {
  sector?: string;
  eligibilityRules: EligibilityRule[];
  whereToApply: string;
  documentsRequired: string[];
  howToApply?: string[];
  sourceUrl?: string;
  lastUpdated: string;
}

export interface BenefitListResponse {
  items: BenefitSummaryAPI[];
  total: number;
  page: number;
  limit: number;
  pages: number;
}

export interface BenefitStatsResponse {
  totalBenefits: number;
  byScope: Record<string, number>;
  byCategory: Record<string, number>;
  statesCovered: number;
  municipalitiesCovered: number;
}

export interface BenefitsByLocationResponse {
  stateCode: string;
  stateName: string;
  municipalityIbge?: string;
  municipalityName?: string;
  federal: BenefitSummaryAPI[];
  state: BenefitSummaryAPI[];
  municipal: BenefitSummaryAPI[];
  sectoral: BenefitSummaryAPI[];
  total: number;
}

// ========== ELIGIBILITY TYPES ==========

export interface RuleEvaluationResult {
  ruleDescription: string;
  passed: boolean;
  inconclusive: boolean;
  field: string;
  expectedValue: unknown;
  actualValue?: unknown;
}

export interface BenefitEligibilityResult {
  benefit: BenefitSummaryAPI;
  status: 'eligible' | 'likely_eligible' | 'maybe' | 'not_eligible' | 'not_applicable' | 'already_receiving';
  matchedRules: string[];
  failedRules: string[];
  inconclusiveRules: string[];
  estimatedValue?: number;
  reason?: string;
  ruleDetails?: RuleEvaluationResult[];
}

export interface EligibilitySummary {
  eligible: BenefitEligibilityResult[];
  likelyEligible: BenefitEligibilityResult[];
  maybe: BenefitEligibilityResult[];
  notEligible: BenefitEligibilityResult[];
  notApplicable: BenefitEligibilityResult[];
  alreadyReceiving: BenefitEligibilityResult[];
  totalAnalyzed: number;
  totalPotentialMonthly: number;
  totalPotentialAnnual: number;
  totalPotentialOneTime: number;
  prioritySteps: string[];
  documentsNeeded: string[];
}

export interface EligibilityResponse {
  profileSummary: Record<string, unknown>;
  summary: EligibilitySummary;
  evaluatedAt: string;
}

export interface EligibilityRequest {
  profile: CitizenProfile;
  scope?: 'federal' | 'state' | 'municipal' | 'sectoral';
  includeNotApplicable?: boolean;
}

// ========== API v2 BASE URL ==========

const V2_BASE = '/v2/benefits';

// ========== BENEFITS ENDPOINTS ==========

/**
 * Fetch paginated list of benefits with optional filters
 */
export async function fetchBenefitsV2(params?: {
  scope?: string;
  state?: string;
  municipalityIbge?: string;
  sector?: string;
  category?: string;
  status?: string;
  search?: string;
  page?: number;
  limit?: number;
}): Promise<BenefitListResponse> {
  const queryParams = new URLSearchParams();

  if (params) {
    Object.entries(params).forEach(([key, value]) => {
      if (value !== undefined && value !== null && value !== '') {
        // Convert camelCase to snake_case for API
        const snakeKey = key.replace(/([A-Z])/g, '_$1').toLowerCase();
        queryParams.append(snakeKey, String(value));
      }
    });
  }

  const queryString = queryParams.toString();
  const url = queryString ? `${V2_BASE}/?${queryString}` : `${V2_BASE}/`;

  const response = await api.get(url);
  return response.data;
}

/**
 * Fetch a single benefit by ID
 */
export async function fetchBenefitByIdV2(benefitId: string): Promise<BenefitDetailAPI> {
  const response = await api.get(`${V2_BASE}/${benefitId}`);
  return response.data;
}

/**
 * Fetch benefits statistics
 */
export async function fetchBenefitsStats(): Promise<BenefitStatsResponse> {
  const response = await api.get(`${V2_BASE}/stats`);
  return response.data;
}

/**
 * Fetch benefits by geographic location
 */
export async function fetchBenefitsByLocation(
  stateCode: string,
  municipalityIbge?: string
): Promise<BenefitsByLocationResponse> {
  const queryParams = municipalityIbge
    ? `?municipality_ibge=${municipalityIbge}`
    : '';

  const response = await api.get(`${V2_BASE}/by-location/${stateCode}${queryParams}`);
  return response.data;
}

// ========== ELIGIBILITY ENDPOINTS ==========

/**
 * Full eligibility check against all benefits
 */
export async function checkEligibilityV2(
  profile: CitizenProfile,
  options?: {
    scope?: 'federal' | 'state' | 'municipal' | 'sectoral';
    includeNotApplicable?: boolean;
  }
): Promise<EligibilityResponse> {
  const request: EligibilityRequest = {
    profile,
    scope: options?.scope,
    includeNotApplicable: options?.includeNotApplicable ?? false,
  };

  const response = await api.post(`${V2_BASE}/eligibility/check`, request);
  return response.data;
}

/**
 * Quick eligibility count (lighter response)
 */
export async function quickEligibilityCheck(
  estado: string,
  rendaFamiliarMensal: number,
  pessoasNaCasa: number,
  cadastradoCadunico: boolean
): Promise<{
  eligibleCount: number;
  maybeCount: number;
  totalPotentialMonthly: number;
  topBenefits: Array<{ name: string; value?: number }>;
}> {
  const params = new URLSearchParams({
    estado,
    renda_familiar_mensal: String(rendaFamiliarMensal),
    pessoas_na_casa: String(pessoasNaCasa),
    cadastrado_cadunico: String(cadastradoCadunico),
  });

  const response = await api.get(`${V2_BASE}/eligibility/quick?${params}`);
  return response.data;
}

// ========== HELPER: CONVERT API TO ENGINE FORMAT ==========

/**
 * Convert API benefit summary to engine Benefit format
 */
export function apiToBenefit(apiBenefit: BenefitDetailAPI): Benefit {
  return {
    id: apiBenefit.id,
    name: apiBenefit.name,
    shortDescription: apiBenefit.shortDescription,
    scope: apiBenefit.scope,
    state: apiBenefit.state,
    municipalityIbge: apiBenefit.municipalityIbge,
    sector: apiBenefit.sector,
    estimatedValue: apiBenefit.estimatedValue,
    eligibilityRules: apiBenefit.eligibilityRules,
    whereToApply: apiBenefit.whereToApply,
    documentsRequired: apiBenefit.documentsRequired,
    howToApply: apiBenefit.howToApply,
    sourceUrl: apiBenefit.sourceUrl,
    lastUpdated: apiBenefit.lastUpdated,
    status: apiBenefit.status as Benefit['status'],
    icon: apiBenefit.icon,
    category: apiBenefit.category,
  };
}

/**
 * Convert API summary to minimal Benefit format (for list views)
 */
export function apiSummaryToBenefit(summary: BenefitSummaryAPI): Partial<Benefit> {
  return {
    id: summary.id,
    name: summary.name,
    shortDescription: summary.shortDescription,
    scope: summary.scope,
    state: summary.state,
    municipalityIbge: summary.municipalityIbge,
    estimatedValue: summary.estimatedValue,
    status: summary.status as Benefit['status'],
    icon: summary.icon,
    category: summary.category,
  };
}

File: backend/app/routers/benefits_v2.py
"""
Benefits API v2 endpoints.
Unified benefits catalog with eligibility evaluation.
"""

from typing import Optional, List

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func

from app.database import get_db
from app.models.benefit import Benefit
from app.schemas.benefit import (
    BenefitResponse,
    BenefitSummary,
    BenefitListResponse,
    BenefitStatsResponse,
    CitizenProfile,
    EligibilityRequest,
    EligibilityResponse,
    EstimatedValue,
)
from app.services.eligibility_service import (
    evaluate_all_benefits,
    get_benefits_for_location,
)

router = APIRouter()


@router.get("/", response_model=BenefitListResponse)
async def list_benefits(
    scope: Optional[str] = Query(None, description="Filter by scope: federal, state, municipal, sectoral"),
    state: Optional[str] = Query(None, description="Filter by state code (e.g., SP, RJ)"),
    municipality_ibge: Optional[str] = Query(None, description="Filter by municipality IBGE code"),
    sector: Optional[str] = Query(None, description="Filter by sector (e.g., pescador, agricultor)"),
    category: Optional[str] = Query(None, description="Filter by category"),
    search: Optional[str] = Query(None, description="Search by name or description"),
    status: Optional[str] = Query("active", description="Filter by status: active, suspended, ended"),
    page: int = Query(1, ge=1, description="Page number"),
    limit: int = Query(50, ge=1, le=200, description="Items per page"),
    db: AsyncSession = Depends(get_db),
):
    """
    List all benefits with optional filters.

    Returns a paginated list of benefits from the unified catalog.
    Includes federal, state, municipal, and sectoral benefits.

    Examples:
    - GET /api/v2/benefits?scope=federal
    - GET /api/v2/benefits?state=SP&scope=state
    - GET /api/v2/benefits?municipality_ibge=3550308
    - GET /api/v2/benefits?search=bolsa
    """
    # Build base query
    stmt = select(Benefit)

    # Apply filters
    if scope:
        stmt = stmt.where(Benefit.scope == scope.lower())

    if state:
        stmt = stmt.where(Benefit.state == state.upper())

    if municipality_ibge:
        stmt = stmt.where(Benefit.municipality_ibge == municipality_ibge)

    if sector:
        stmt = stmt.where(Benefit.sector == sector.lower())

    if category:
        stmt = stmt.where(Benefit.category.ilike(f"%{category}%"))

    if status:
        stmt = stmt.where(Benefit.status == status.lower())

    if search:
        search_filter = f"%{search.lower()}%"
        stmt = stmt.where(
            (Benefit.name.ilike(search_filter)) |
            (Benefit.short_description.ilike(search_filter))
        )

    # Count total
    count_stmt = select(func.count()).select_from(stmt.subquery())
    count_result = await db.execute(count_stmt)
    total = count_result.scalar() or 0

    # Paginate
    offset = (page - 1) * limit
    stmt = stmt.order_by(Benefit.name).offset(offset).limit(limit)
    result = await db.execute(stmt)
    benefits = result.scalars().all()

    # Convert to response format
    items = []
    for b in benefits:
        est_value = None
        if b.estimated_value:
            est_value = EstimatedValue(
                type=b.estimated_value.get("type", "monthly"),
                min=b.estimated_value.get("min"),
                max=b.estimated_value.get("max"),
                description=b.estimated_value.get("description"),
            )
        items.append(BenefitSummary(
            id=b.id,
            name=b.name,
            shortDescription=b.short_description,
            scope=b.scope,
            state=b.state,
            municipalityIbge=b.municipality_ibge,
            estimatedValue=est_value,
            status=b.status,
            icon=b.icon,
            category=b.category,
        ))

    return BenefitListResponse(
        items=items,
        total=total,
        page=page,
        limit=limit,
        pages=(total + limit - 1) // limit if limit > 0 else 0,
    )


@router.get("/stats", response_model=BenefitStatsResponse)
async def get_benefits_stats(
    db: AsyncSession = Depends(get_db),
):
    """
    Get statistics about the benefits catalog.

    Returns total counts by scope, category, and geographic coverage.
    """
    # Total benefits
    total_stmt = select(func.count()).select_from(Benefit)
    total_result = await db.execute(total_stmt)
    total = total_result.scalar() or 0

    # By scope
    scope_stmt = select(
        Benefit.scope,
        func.count().label("count")
    ).group_by(Benefit.scope)
    scope_result = await db.execute(scope_stmt)
    by_scope = {row.scope: row.count for row in scope_result}

    # By category
    category_stmt = select(
        Benefit.category,
        func.count().label("count")
    ).where(Benefit.category.isnot(None)).group_by(Benefit.category)
    category_result = await db.execute(category_stmt)
    by_category = {row.category: row.count for row in category_result}

    # States covered
    states_stmt = select(func.count(func.distinct(Benefit.state))).where(
        Benefit.state.isnot(None)
    )
    states_result = await db.execute(states_stmt)
    states_covered = states_result.scalar() or 0

    # Municipalities covered
    muni_stmt = select(func.count(func.distinct(Benefit.municipality_ibge))).where(
        Benefit.municipality_ibge.isnot(None)
    )
    muni_result = await db.execute(muni_stmt)
    municipalities_covered = muni_result.scalar() or 0

    return BenefitStatsResponse(
        totalBenefits=total,
        byScope=by_scope,
        byCategory=by_category,
        statesCovered=states_covered,
        municipalitiesCovered=municipalities_covered,
    )


@router.get("/by-location/{state_code}")
async def get_benefits_by_location(
    state_code: str,
    municipality_ibge: Optional[str] = Query(None, description="IBGE code for municipal benefits"),
    db: AsyncSession = Depends(get_db),
):
    """
    Get all applicable benefits for a specific location.

    Returns federal + state + municipal benefits for the given location.

    Examples:
    - GET /api/v2/benefits/by-location/SP (federal + SP state benefits)
    - GET /api/v2/benefits/by-location/SP?municipality_ibge=3550308 (includes São Paulo city benefits)
    """
    benefits = await get_benefits_for_location(
        db,
        state_code=state_code.upper(),
        ibge_code=municipality_ibge,
    )

    # Group by scope
    federal = []
    state = []
    municipal = []
    sectoral = []

    for b in benefits:
        item = b.to_dict()
        if b.scope == "federal":
            federal.append(item)
        elif b.scope == "state":
            state.append(item)
        elif b.scope == "municipal":
            municipal.append(item)
        elif b.scope == "sectoral":
            sectoral.append(item)

    return {
        "state": state_code.upper(),
        "municipality_ibge": municipality_ibge,
        "total": len(benefits),
        "federal": federal,
        "state": state,
        "municipal": municipal,
        "sectoral": sectoral,
    }


@router.get("/{benefit_id}")
async def get_benefit(
    benefit_id: str,
    db: AsyncSession = Depends(get_db),
):
    """
    Get full details of a specific benefit.

    Examples:
    - GET /api/v2/benefits/federal-bolsa-familia
    - GET /api/v2/benefits/sp-bolsa-povo
    """
    stmt = select(Benefit).where(Benefit.id == benefit_id)
    result = await db.execute(stmt)
    benefit = result.scalar_one_or_none()

    if not benefit:
        raise HTTPException(status_code=404, detail="Benefício não encontrado")

    return benefit.to_dict()


@router.post("/eligibility/check", response_model=EligibilityResponse)
async def check_eligibility(
    request: EligibilityRequest,
    db: AsyncSession = Depends(get_db),
):
    """
    Evaluate eligibility for all benefits based on citizen profile.

    Receives a CitizenProfile and returns eligibility status for all
    applicable benefits, grouped by status (eligible, likely_eligible,
    maybe, not_eligible, not_applicable, already_receiving).

    Also returns:
    - Total potential monthly/annual/one-time values
    - Priority steps (what to do first)
    - Documents needed

    Example request:
    ```json
    {
      "profile": {
        "estado": "SP",
        "municipioIbge": "3550308",
        "pessoasNaCasa": 4,
        "rendaFamiliarMensal": 800,
        "cadastradoCadunico": true
      }
    }
    ```
    """
    response = await evaluate_all_benefits(
        db=db,
        profile=request.profile,
        scope=request.scope,
        include_not_applicable=request.include_not_applicable,
    )

    return response


@router.post("/eligibility/quick")
async def quick_eligibility_check(
    estado: str = Query(..., description="UF code (e.g., SP)"),
    renda_familiar: float = Query(..., ge=0, description="Monthly family income"),
    pessoas_na_casa: int = Query(1, ge=1, description="Number of people in household"),
    cadastrado_cadunico: bool = Query(False, description="Registered in CadÚnico?"),
    db: AsyncSession = Depends(get_db),
):
    """
    Quick eligibility check with minimal parameters.

    Useful for a simplified initial assessment.
    For full evaluation, use POST /eligibility/check.
    """
    profile = CitizenProfile(
        estado=estado.upper(),
        pessoasNaCasa=pessoas_na_casa,
        rendaFamiliarMensal=renda_familiar,
        cadastradoCadunico=cadastrado_cadunico,
        quantidadeFilhos=0,
        temIdoso65Mais=False,
        temGestante=False,
        temPcd=False,
        temCrianca0a6=False,
        trabalhoFormal=False,
        temCasaPropria=False,
        moradiaZonaRural=False,
        recebeBolsaFamilia=False,
        recebeBpc=False,
        temMei=False,
        trabalhaAplicativo=False,
        agricultorFamiliar=False,
        pescadorArtesanal=False,
        catadorReciclavel=False,
        estudante=False,
        redePublica=False,
    )

    request = EligibilityRequest(
        profile=profile,
        scope=None,
        includeNotApplicable=False,
    )

    response = await evaluate_all_benefits(
        db=db,
        profile=profile,
        scope=None,
        include_not_applicable=False,
    )

    # Return simplified response
    return {
        "estado": estado.upper(),
        "rendaPerCapita": renda_familiar / pessoas_na_casa,
        "totalEligible": len(response.summary.eligible),
        "totalLikelyEligible": len(response.summary.likely_eligible),
        "totalPotentialMonthly": response.summary.total_potential_monthly,
        "topBenefits": [
            {
                "id": r.benefit.id,
                "name": r.benefit.name,
                "estimatedValue": r.estimated_value,
            }
            for r in (response.summary.eligible + response.summary.likely_eligible)[:5]
        ],
        "nextStep": response.summary.priority_steps[0] if response.summary.priority_steps else None,
    }
